torch/utils/data/utils.py

```diff
@@ -0,0 +1,71 @@
+import sys
+import traceback
+
+if sys.version_info[0] == 2:
+    import Queue as queue
+else:
+    import queue
+
+
+class ExceptionWrapper(object):
+    r"Wraps an exception plus traceback to communicate across threads"
+
+    def __init__(self, exc_info):
+        self.exc_type = exc_info[0]
+        self.exc_msg = "".join(traceback.format_exception(*exc_info))
+
+
+# Queue get and put methods may fail if the underlying syscalls are interrupted
+# (EINTR). So we need to wrap them to retry automatically if needed.
+#
+# Syscalls are automatically retried upon encountering EINTR since Python 3.5
+# https://www.python.org/dev/peps/pep-0475/
+# EINTR is not available on Windows.
+if sys.platform == 'win32' or sys.version_info >= (3, 5):
+    def QueueWrapper(queue):
+        return queue
+else:
+    import time
+    import errno
+
+    if sys.version_info >= (3, 3):
+        time_fn = time.perf_counter
+    else:
+        time_fn = time.time
+
+    class QueueWrapper(object):
+        r"""Wraps a queue object that conforms to following interface:
+
+            .get()  or  .get(timeout=timeout)
+            .put(value)
+        """
+
+        def __init__(self, queue):
+            self.queue = queue
+
+        def get(self, timeout=None):
+            while True:
+                try:
+                    if timeout is None:
+                        return self.queue.get()
+                    else:
+                        t0 = time_fn()
+                        return self.queue.get(timeout=timeout)
+                except IOError as e:
+                    if e.errno != errno.EINTR:
+                        raise
+                    if timeout is not None:
+                        timeout -= time_fn() - t0
+                        if timeout <= 0:
+                            raise queue.Empty
+
+        def put(self, val):
+            while True:
+                try:
+                    return self.queue.put(val)
```

Comment:
One thing the comes mind is what happens if EINTR comes in the middle of writing to a queue? This will mean that the data has been corrupted and we can't really do anything about it, right? Similarly with `get()` (what if we get interrupted in the middle of reading!)

Reply:
According to manual, `recv` can get EINTR when it is interrupted before reading any data, and `send` can get EINTR before any data is transmitted.

Reply:
Aren't queues implemented around pipes? If so they are using `read` and `write`. What I'm concerned about is that if I were writing such a queue I would do it like this:
```cpp
std::size_t remaining_bytes = ...;
while (remaining_bytes > 0) {
  ssize_t read_bytes = read(fd, &buffer, remaining_bytes);
  if (read_bytes == -1) throw some_error();
  remaining_bytes -= read_bytes;
  buffer += read_bytes;
}
```
This is because `read` is not guaranteed to return you exactly `remaining_bytes`, but *at most* `remaining_bytes` so you might need to loop. Now, if Python implements read in this way, but breaks out of the loop if it sees EINTR, then by the time we call `get()` again, some of the data could have still been left in the pipe, and it will be corrupted.

Reply:
You are right. I checked python implementation. Queues are implemented using`read` and `write`, and this is indeed problematic. In this case, I can't think of a clean solution currently. Do you have any ideas?

Reply:
Well if EINTR makes the data inside the queue corrupted, then I don't think we have a better choice than crashing 😕

Reply:
Yeah, I'll add a comment about it then.

Reply:
Well I'm just not sure if we still want to merge this patch, since it's likely to crash on the next `get()` anyway

Reply:
Hmm yeah. I'll remove the EINTR part then :) Other parts of the patch should still be reasonable.

There is another issue that I can add a quick fix in that patch. Some people have been using dataloder in a separate thread. Then because the first `DataLoaderIter` object is init'd in a child thread, setting SIGCHLD handler will error at `signal.signal` complaining that it should only be used in main thread. There are two solutions: 1) setting handler at import time or 2) try catch the `signal.signal` (or test if it is main thread using `isinstance(threading.current_thread(), threading._MainThread)`). I am inclined towards the second because the first will still fail if import is done in the multithreading function. Does it sound reasonable to you?